### ISSUE TITLE
Improve auto-completion of filenames

### ISSUE THREAD

**Issue by benjaminweb** [fish-shell/fish-shell#2547]
There are file1.py and file1_test.py only.

``hg add file1.py file1…``

There, fish could see that file1.py has been provided earlier.
As it makes no sense to repeat the same file name, fish could auto-suggest file1_test.py.

**Comment by eush77** [comment]
[Counter-example](http://stackoverflow.com/a/13038235/2424184):

```fish
$ seq 1 5 >one
$ seq 1 2 10 >two
$ cat one two two |sort |uniq -u
2
4
```

**Comment by faho** [comment]
Yes, this isn't something that could be enabled for all completions - there would need to be a way to opt-out or opt-in (I think opt-out might be okay for file-completion).

It also can't be done for _all kinds_ of completions in general, since e.g. options might be specified multiple times (rather common for "-v" or "--verbose" options).

Doing it correctly would also be suprisingly hard - since filenames are almost completely unrestricted on UNIXy systems, and since there's quite a variety of cases in which something that looks like a filename may appear in a different context. Long story short: This would pretty much require something like #478.

For suggestions, however, which don't need to be 100% correct - only very likely to be, I think it'd be okay to enable, even without an option to disable.

**Comment by pickfire** [comment]
Yeah, I would like this features sometimes althought it could cost a bit of performance.

I would be nice to just **ignore the repeated pattern**.

**Comment by benjaminweb** [comment]
+1
